### PR TITLE
Improve the usability of the ca app using EdDSA

### PR DESCRIPTION
**Pull request by mattcaswell**
Previously you had to supply "null" as the digest to use EdDSA. This changes
things so that any digest is ignored.

### CODE REVIEW

File: apps/ca.c [hunk 1]
```diff
@@ -728,24 +728,28 @@ int ca_main(int argc, char **argv)
...
-    if (md == NULL && (md = lookup_conf(conf, section, ENV_DEFAULT_MD)) == NULL)
-        goto end;
-
-    if (strcmp(md, "null") == 0) {
+    def_ret = EVP_PKEY_get_default_digest_nid(pkey, &def_nid);
+    /*
+     * EVP_PKEY_get_default_digest_nid() returns 2 if the digest is
+     * mandatory for this algorithm.
+     */
+    if (def_ret == 2 && def_nid == NID_undef) {
```
Comment (richsalz) [inline apps/ca.c]: dislike magic numbers, but okay if this is out of scope for this PR.

Comment (mattcaswell) [reply apps/ca.c]: So do I, but I thought it was moving the scope too far to tackle that in this PR.

**Comment by romen** [comment]
I guess this is perfectly fine for now. 

I only want to raise the issue of thinking about a future when instead of only PureEdDSA OpenSSL might support also HashEdDSA: would this change have an unintended impact on configuration patterns or command line arguments that might be difficult to break in future versions?

I honestly don't know the answer, nor if such issues would concern the guarantees in minor or major releases.

**Comment by mattcaswell** [comment]
> I only want to raise the issue of thinking about a future when instead of only PureEdDSA OpenSSL might support also HashEdDSA: would this change have an unintended impact on configuration patterns or command line arguments that might be difficult to break in future versions?

I did consider this, but I don't think it is an issue. The way it has been implemented the ca app asks the EVP_PKEY object what its default hash is. Only if the default hash is undefined is it allowed to not have a hash specified. If we later implement hash EdDSA then the EVP_PKEY will return with a sensible, defined, hash.

**Comment by romen** [comment]
Thanks for clarifying it @mattcaswell!

**Comment by beldmit** [comment]
It seems to me that req.c be fixed the similar way to ca.c.

**Comment by petrovr** [comment]
I'm not sure that code should ignore configuration settings. Correction for #6201 , i.e. null digest is fine with me.

**Comment by mattcaswell** [comment]
> I'm not sure that code should ignore configuration settings. Correction for #6201 , i.e. null digest is fine with me

The current solution is causing some significant usability problems, so it doesn't seem to be viable:

https://mta.openssl.org/pipermail/openssl-users/2018-August/008473.html

**Comment by mattcaswell** [comment]
I added a new commit to make an equivalent change in the req app. Interestingly, while updating the documentation for that I discovered that we already ignore the digest configuration setting in req in some situations so this approach seems consistent with that.

**Comment by mattcaswell** [comment]
@richsalz please reconfirm

**Comment by richsalz** [comment]
reconfirm.  good catch, @petrovr !

**Comment by mattcaswell** [comment]
I think the catch was @beldmit...!

**Comment by petrovr** [comment]
mattcaswell wrote:
> I think the catch was @beldmit...!
Perhaps  @beldmit is reporter.

My point is that program code can not ignore invalid values:
1) null is valid value,
2) missing digest could mean use default , i.e. null
3) other values, for instance sha1, should break processing .

Regards,
Roumen Petrov

**Comment by mattcaswell** [comment]
> My point is that program code can not ignore invalid values:
>1) null is valid value,
>2) missing digest could mean use default , i.e. null
>3) other values, for instance sha1, should break processing 

I disagree. It is perfectly reasonable to want to set a default digest of (say) sha256 for most stuff, but not use it when doing EdDSA.

**Comment by petrovr** [comment]
mattcaswell wrote:
>> My point is that program code can not ignore invalid values:
>> 1) null is valid value,
>> 2) missing digest could mean use default , i.e. null
>> 3) other values, for instance sha1, should break processing
> 
> I disagree. It is perfectly reasonable to want to set a default digest of (say) sha256 for most stuff, but not use it when doing EdDSA.

Digest depends from algorithm.
I would like to have more consistent behaviour. For instance if I set 
md5 for dsa processing is  blocked by error  ..."dsa 
routines:pkey_dsa_ctrl:invalid digest type:crypto/dsa/dsa_pmeth.c:142:"
This is reason to ask for consistency with existing functionality.

Roumen

**Comment by mattcaswell** [comment]
I have the required approval for this, and I still think this is the correct approach, but given @petrovr's objection I invite any other comment from @openssl/committers.

**Comment by t-j-h** [comment]
I don't see the parameter as forcing a digest value conceptually.

I think defining default_digest to mean the digest for algorithms that allow a settable digest is a reasonable approach (and the documentation should be updated to make that clear that this is the definition of the configuration parameter - not that some algorithms ignore it - that it only applies to algorithms that allow settable values). That way it is clearer what is going on here and looks less like a workaround.

Basically more clarity in the docs wording - but I agree with @mattcaswell in terms of the approach.

**Comment by petrovr** [comment]
It is about how to change existing code :
...
    if (md == NULL && (md = lookup_conf(conf, section, ENV_DEFAULT_MD)) == NULL)
        goto end;
...
, i.e. parameter "default_md" has to be specified either on command line or in configuration section.
Above could be split as follow:
...
  if (md == null) md = lookup_conf(...)
  if (md == null) {
    if md is required : goto end
    else md = "default" 
  }
...
I expect above code to allow user to use default_md=default and with proposed modifications (correction in existing function and addition of new one) for ASN1_PKEY_CTRL_DEFAULT_MD_NID "null" digest to be used.

Also think about "default" to be set unconditionally in above code snippet.

**Comment by petrovr** [comment]
I don't like to stop progress on this issue, i.e. approvals.

For protocol I did tests with following code snipped from apps/ca.c
-----
    if (md == NULL)
        md = lookup_conf(conf, section, ENV_DEFAULT_MD);

    if (md == NULL)
        md = "default";

    if (strcmp(md, "null") == 0) {
        dgst = EVP_md_null();
    } else {
        if (strcmp(md, "default") == 0) {
            int def_nid;
            if (EVP_PKEY_get_default_digest_nid(pkey, &def_nid) <= 0) {
                BIO_puts(bio_err, "no default digest\n");
                goto end;
            }
            /* NID_undef if signing algorithm requires there to be no digest */
            dgst = (def_nid == NID_undef)
                ? EVP_md_null()
                : EVP_get_digestbynid(def_nid);
        }
    }
    if (dgst == NULL) {
        if (!opt_md(md, &dgst))
            goto end;
    }
-----
This was my idea. Let me know if someone found above generic approach useful. Is so I open an enhancement request.
In all cases go ahead with current status.

File: apps/ca.c (hunk 1 above)
Comment (romen) [reply apps/ca.c]: On an additional note: we say that if it returns 2 the digest is mandatory for this algorithm, but we act in that direction only if the return is 2 and the required digest is `NID_undef`.

What if the required default was something else, e.g. SM3 for SM2 or an engine-provided PKEY method with such requirements? Shouldn't it be honored?

File: apps/req.c [hunk 2]
```diff
@@ -1601,10 +1601,19 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
...
+    int i, def_nid;
 
     if (ctx == NULL)
         return 0;
+    /*
+     * EVP_PKEY_get_default_digest_nid() returns 2 if the digest is mandatory
+     * for this algorithm.
+     */
+    if (EVP_PKEY_get_default_digest_nid(pkey, &def_nid) == 2
+            && def_nid == NID_undef) {
```
Comment (romen) [inline apps/req.c]: Also on `req` we honor the return value 2 only for `NID_undef`

File: doc/man1/req.pod [hunk 3]
```diff
@@ -394,10 +394,10 @@ option. For compatibility B<encrypt_rsa_key> is an equivalent option.
...
 =item B<default_md>
 
-This option specifies the digest algorithm to use.
-Any digest supported by the OpenSSL B<dgst> command can be used.
-If not present then MD5 is used.
-This option can be overridden on the command line.
+This option specifies the digest algorithm to use. Any digest supported by the
+OpenSSL B<dgst> command can be used. This option can be overridden on the
+command line. Certain signing algorithms (i.e. Ed25519 and Ed448) will ignore
+any digest that has been set.
```
Comment (romen) [inline doc/man1/req.pod]: This is a behavior change, should it warrant an HISTORY entry?

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod [hunk 4]
```diff
@@ -13,7 +13,8 @@ EVP_PKEY_get_default_digest_nid - get default signature digest
 
 The EVP_PKEY_get_default_digest_nid() function sets B<pnid> to the default
 message digest NID for the public key signature operations associated with key
-B<pkey>.
+B<pkey>. Note that some signature algorithms (i.e. Ed25519 and Ed448) do not use
+a digest during signing. In this case B<pnid> will be set to NID_undef.
```
Comment (romen) [inline doc/man3/EVP_PKEY_get_default_digest_nid.pod]: If I am not mistaken, up to 1.1.0 returning `NID_undef` as the default md would cause errors in the OpenSSL apps.

If that's the case, and this change is ok for a minor release, an HISTORY entry detailing the change might be required.

Comment (romen) [reply doc/man3/EVP_PKEY_get_default_digest_nid.pod]: On further investigation, I am mistaken on this, so nevermind this comment.

**Comment by beldmit** [comment]
BTW, why not to specify a separate NID, not NID_undef for md_null? It can save us from separate processing NID_undef as dgst_nid here and may be in some other places of code.

**Comment by kaduk** [comment]
> BTW, why not to specify a separate NID, not NID_undef for md_null? It can save us from separate processing NID_undef as dgst_nid here and may be in some other places of code.

One could perhaps argue that the "Intrinsic" TLS HashAlgorithm defined by RFC 8422 (section 5.1.3) would argue for having such a dedicated new NID, but if we do add such a thing, we'd need to assign it some semantics for all the other places it could be used, and it's unclear that we really want to do that.

File: apps/ca.c (hunk 1 above)
Comment (mattcaswell) [reply apps/ca.c]: @romen, perhaps it should - but again I think that starts moving the scope of this PR too far away from the original problem.

File: apps/req.c (hunk 2 above)
Comment (mattcaswell) [reply apps/req.c]: As above I think this is a valid point, but I don't want to tackle that in this PR.

File: doc/man1/req.pod (hunk 3 above)
Comment (mattcaswell) [reply doc/man1/req.pod]: Is it really a behaviour change? There aren't any algorithms in 1.1.0 where this clause applies. You gave the example somewhere else of someone providing an engine to do ed25519 in 1.1.0. I'm not convinced that is actually possible. In 1.1.0 we assume that all signing algorithms require a digest...and stuff will probably break if you try to supply one via an engine where that isn't true.

File: apps/ca.c (hunk 1 above)
Comment (romen) [reply apps/ca.c]: @mattcaswell I understand that.

My issue is that I'm just not convinced that, in this case, usability can trample on backwards behavioral compatibility.

Is it really necessary that Ed25519/448 should ignore explicit configuration lines/cmdline args?
I ask because to introduce this we are adding a very specific (only for `(EVP_PKEY_get_default_digest_nid() == 2) && (returned_nid == NID_undef)` behavior in the apps, which does not match exactly the documented behavior of the function (*returns 2 if the digest is mandatory for this algorithm*).

Comment (mattcaswell) [reply apps/ca.c]: I'm struggling to see this as a break in backwards compat. If `EVP_PKEY_get_default_digest_nid()` returns 2 then it means the digest is mandatory. If the NID for that mandatory digest is `NID_undef` then what does that mean under the current documentation? About the best one can say is that that behaviour is "undefined". And in practice it makes no difference, because until 1.1.1 came along that never happens! So why does this break backwards compat?

File: doc/man1/req.pod (hunk 3 above)
Comment (romen) [reply doc/man1/req.pod]: I commented on the behavior change explicitly because I noticed that it started breaking stuff for an engine I am working on ([libsuola](https://github.com/romen/libsuola)) which does exactly this, and I noticed that after this change 1.1.1 apps have the option of ignoring explicitly set configuration entries or cmdline args, while 1.1.0 apps don't.

I just wanted to make sure that all the implications of this change are discussed before committing to them! ;)

File: apps/ca.c (hunk 1 above)
Comment (romen) [reply apps/ca.c]: > So why does this break backwards compat?

Because if you run the same commands with 1.1.0 apps (with access to Ed25519 through an engine) and 1.1.1 apps, using the same configuration file for the pki settings, the former fails while the second succeeds.

File: doc/man1/req.pod (hunk 3 above)
Comment (mattcaswell) [reply doc/man1/req.pod]: > I noticed that it started breaking stuff for an engine I am working on

Interesting. Can you be more explicit about in what way it broke?

File: apps/ca.c (hunk 1 above)
Comment (mattcaswell) [reply apps/ca.c]: Is that a break in backwards compat? There are lots of things that the apps can now do in 1.1.1 that they couldn't do before. Try pointing a 1.1.0 s_client at a TLSv1.3 only server....it will fail. Do the same in 1.1.1 and it will work!

Comment (mattcaswell) [reply apps/ca.c]: More concerning for backwards compat would be if there are things in 1.1.0 that used to work, but now don't.

Comment (vdukhovni) [reply apps/ca.c]: > @romen Because if you run the same commands with 1.1.0 apps (with access to Ed25519 through an engine) and 1.1.1 apps, using the same configuration file for the pki settings, the former fails while the second succeeds.

The whole point of the PR is to succeed where we previously failed. The failure is spurious, the public key algorithm does not use a digest for signing, so we should succeed regardless of any specified digest. Requiring the user to explicitly specify a null digest is a bad UI and breaks existing configuration files and command-line scripts.

If a particular public key algorithm works exclusively with a particular digest or no digest, then we should always use that (or no) digest, and ignore options that provide a digest to use for those key types that could use one of many.

@mattcaswell I think it is time to go ahead and push this PR, I've not seen any substantive objections that prevent moving forward. If someone wants to fine-tune with a follow-on PR we can evaluate that separately.

Comment (romen) [reply apps/ca.c]: > If a particular public key algorithm works exclusively with a particular digest or no digest, then we should always use that (or no) digest, and ignore options that provide a digest to use for those key types that could use one of many.

@vdukhovni but we don't do this with this changes, only for the "no digest", that's why I am talking about "special case" and not conforming to the documented behavior: if the function returns 2 but it's not asking for the null digest then we still fail!

@mattcaswell I don't mean to stop the PR, I raised my objections and discussion points, which were evaluated, so I am satisfied anyway!

**Comment by mattcaswell** [comment]
For reasons I don't entirely understand (or agree with), this PR seems to have been more controversial than I expected. However I still believe this is the correct approach, and it has all the relevant approvals and support from an OMC member, with no "-1" from any committer, so I have pushed. If people want to they can follow up with further PRs as appropriate.

File: apps/ca.c (hunk 1 above)
Comment (vdukhovni) [reply apps/ca.c]: I am aware that this PR does not address the case of a single non-null mandatory digest.  We can address that issue when we come to it.  This PR is strictly an improvement, that addresses the immediate issue.  We can generalize later, but feel free to propose a follow-on PR if you have something ready to go.
